### ISSUE TITLE
Change to vertical indexing for W field in MOM5 B-grid output

### ISSUE THREAD

**Issue by hrsdawson** [OceanParcels/parcels#922]
Hi @erikvansebille and @daanreijnders. As discussed, the vertical velocity in MOM5 is defined at the base of the tracer cell, shown in the sketch below. Parcels currently expects the vertical velocity to be defined at the top surface of the grid cell, meaning that `w` is applied at the upper face rather than the bottom face of the grid cell when using MOM5 output in Parcels. Could the vertical indexing be modified for this case so that `w` is applied correctly to the base of the grid cell? This would be a neater approach than PR #890. Modifying the vertical indexing for `w` in this way assists with, though does not fix, the issue of stuck particles along coastal boundaries. 

[Indexing_sketch.pdf](https://github.com/OceanParcels/parcels/files/5236795/Indexing_sketch.pdf)

It does however raise the question of what `w` values to use at the surface boundary. I am currently applying `w` from the bottom of the top z layer (using the stationary z layer approximation) as discussed in issue #860. This causes particles to move through the surface which then need to be mapped to a submerge kernel. The other option is to set `w` to zero at the surface of the top z layer which would prevent the through-surface issue. I'm not sure which option is more appropriate - @Adele157 do you have any more thoughts on this?  

I have included a unit test for the vertical indexing below (apologies, couldn't get it to format properly). This is based on the MITgcm grid unit test with indexing modified in the vertical rather than the horizontal. I hope it works. I will send some time snapshots of the hydrodynamic data to you shortly. 

It would also be great if this [tutorial ](https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/documentation_indexing.ipynb) could be updated to include some more information on B-grids, such as MOM5 and this vertical indexing issue. I think this could be beneficial to other/future users. 

```python
@pytest.mark.parametrize('mode', ['scipy', 'jit'])
@pytest.mark.parametrize('gridindexingtype', ['mom5', 'nemo'])
@pytest.mark.parametrize('withtime', [False, True])`
def test_mom5gridindexing_3D(mode, gridindexingtype, withtime):
    xdim = zdim = 201
    ydim = 2
    a = c = 20000  # domain size
    b = 2
    lon = np.linspace(-a / 2, a / 2, xdim, dtype=np.float32)
    lat = np.linspace(-b / 2, b / 2, ydim, dtype=np.float32)
    depth = np.linspace(-c / 2, c / 2, zdim, dtype=np.float32)
    dx, dz = lon[1] - lon[0], depth[1] - depth[0]
    omega = 2 * np.pi / delta(days=1).total_seconds()
    if withtime:
        time = np.linspace(0, 24*60*60, 10)
        dimensions = {"lon": lon, "lat": lat, "depth": depth, "time": time}
        dsize = (time.size, depth.size, lat.size, lon.size)
   else:
       dimensions = {"lon": lon, "lat": lat, "depth": depth}
       dsize = (depth.size, lat.size, lon.size)

    index_signs = {'nemo': 0, 'mom5': 1}
    isign = index_signs[gridindexingtype]

    def calc_r_phi(ln, dp):
        return np.sqrt(ln ** 2 + dp ** 2), np.arctan2(ln, dp)

    def calculate_UVWR(lat, lon, depth, dx, dz, omega):
        U = np.zeros(dsize, dtype=np.float32)
        V = np.zeros(dsize, dtype=np.float32)
        W = np.zeros(dsize, dtype=np.float32)
        R = np.zeros(dsize, dtype=np.float32)

        for i in range(lon.size):
            for j in range(lat.size):
                for k in range(depth.size):
                    r, phi = calc_r_phi(lon[i], depth[k])
                    if withtime:
                        R[:, k, j, i] = r
                    else:
                        R[k, j, i] = r
                    r, phi = calc_r_phi(lon[i] - dx / 2, depth[k] + isign * dz)
                    if withtime:
                        W[:, k, j, i] = -omega * r * np.sin(phi)
                    else:
                        W[k, j, i] = -omega * r * np.sin(phi)
                    r, phi = calc_r_phi(lon[i], depth[k] + dz / 2)
                    if withtime:
                        U[:, k, j, i] = omega * r * np.cos(phi)
                    else:
                        U[k, j, i] = omega * r * np.cos(phi)
        return U, V, W, R

    U, V, W, R = calculate_UVWR(lat, lon, depth, dx, dz, omega)
    data = {"U": U, "V": V, "W": W, "R": R}
    fieldset = FieldSet.from_data(data, dimensions, mesh="flat")
    if gridindexingtype == 'mom5':
        fieldset.U.interp_method = "bgrid_velocity"
        fieldset.V.interp_method = "bgrid_velocity"
        fieldset.W.interp_method = "bgrid_w_velocity"
    else:
        fieldset.U.interp_method = "cgrid_velocity"
        fieldset.V.interp_method = "cgrid_velocity"
        fieldset.W.interp_method = "cgrid_velocity"
    fieldset.U.gridindexingtype = gridindexingtype
    fieldset.V.gridindexingtype = gridindexingtype
    fieldset.W.gridindexingtype = gridindexingtype

    def UpdateR(particle, fieldset, time):
        particle.radius = fieldset.R[time, particle.depth, particle.lat, particle.lon]

    class MyParticle(ptype[mode]):
        radius = Variable('radius', dtype=np.float32, initial=0.)
        radius_start = Variable('radius_start', dtype=np.float32, initial=fieldset.R)

    pset = ParticleSet(fieldset, pclass=MyParticle, depth=4e3, lon=0, lat=0, time=0)

    pset.execute(pset.Kernel(UpdateR) + AdvectionRK4_3D,
                 runtime=delta(hours=14), dt=delta(minutes=5))
    assert np.allclose(pset.radius, pset.radius_start, atol=10)
```

**Comment by erikvansebille** [comment]
Thanks for this code and especially the sketch, @hrsdawson! This is a really useful starting point for us to implement the MOM5 grid indexing. I've just created a unit test at #923 (folded in with the other unit tests) based on your code above (which I've formatted for you using the code-block mark-down syntax with three `)

The problem now is that the unit test passes, even though we have not changed anything to the actual interpolation routines yet. That is very strange, and I don't see how it could be happening. I'll explore further next week..
